src/PanoImageRenderer/VideoLoader.js

```diff
@@ -40,45 +68,88 @@ export default class VideoLoader {
 		videoType && (sourceElement.type = videoType);
 
 		this._video.appendChild(sourceElement);
-		this._sourceCount++;
 		return true;
 	}
 
 	set(video) {
 		this._reset(); // reset resources.
 
+		if (!video) {
+			return;
+		}
+
 		if (video instanceof HTMLVideoElement) {
 			// video tag
 			this._video = video;
 		} else if (typeof video === "string" || typeof video === "object") {
 			// url
 			this._video = document.createElement("video");
-			this._video.crossOrigin = "anonymous";
+			this._video.setAttribute("crossorigin", "anonymous");
+			this._video.setAttribute("webkit-playsinline", "");
+			this._video.setAttribute("playsinline", "");
 
 			if (video instanceof Array) {
 				video.forEach(v => this._appendSourceElement(v));
 			} else {
 				this._appendSourceElement(video);
 			}
+		}
+
+		if (this._video) {
+			// count sources to count
+			if (!this._video.getAttribute("src")) {
+				this._sourceCount = this._video.querySelectorAll("source").length;
+			} else {
+				this._sourceCount = 1;
+			}
 
 			if (this._sourceCount > 0) {
-				this._video.load();
+				if (this._video.readyState < this._thresholdReadyState) {
+					this._video.load();
+					// attach loading error listener
+					this._attachErrorHandler(this._onerror);
+				}
 			} else {
 				this._video = null;
 			}
 		}
 	}
 
+	_attachErrorHandler(handler) {
+		this._video.addEventListener("error", handler);
+		this._sources = this._video.querySelectorAll("source");
+		[].forEach.call(this._sources, source => {
+			source.addEventListener("error", handler);
+		});
+	}
+
+	_detachErrorHandler(handler) {
+		this._video.removeEventListener("error", handler);
+		[].forEach.call(this._sources, source => {
+			source.removeEventListener("error", handler);
+		});
+	}
+
 	get() {
 		return new Promise((res, rej) => {
 			if (!this._video) {
 				rej("VideoLoader: video is undefined");
-			} else if (this._video.readyState >= READY_STATUS.HAVE_CURRENT_DATA) {
+			} else if (this._loadStatus === READY_STATUS.LOADING_FAILED) {
+				rej("VideoLoader: video source does not exist");
```

Comment:
@happyhj I think `video source is invalid.` is more proper message. How about you?

Reply:
I also prefer `video source is invalid` because it is not a definite expression.